File: src/main/java/com/mmust/token/BaseToken.java
package com.mmust.token;

abstract class BaseToken implements Token {

	private String value;

	BaseToken(String value) {
		this.value = value;
	}

	@Override
	public String getRawValue() {
		return value;
	}

	@Override
	public String toString() {
		return value;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;

		BaseToken baseToken = (BaseToken) o;

		return value.equals(baseToken.value);
	}

	@Override
	public int hashCode() {
		return value.hashCode();
	}
}

File: src/main/kotlin/com/mmust/token/BaseToken.kt
package com.mmust.token

internal abstract class BaseToken internal constructor(override val rawValue: String) : Token {

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as BaseToken

        if (rawValue != other.rawValue) return false

        return true
    }

    override fun hashCode(): Int {
        return rawValue.hashCode()
    }

    override fun toString() = rawValue
}
